### DISCUSSION TITLE
Promoting Next.js app on Heroku without rebuilding

### DISCUSSION THREAD

**Discussion by patjakubik** [vercel/next.js#20068] [Help] [answered]
Env variables in Next.js apps are set during build time, that means that when promoting app on Heroku from dev to staging, and from staging to prod, it will always use ENV variables from dev. In the past, I used publicRuntimeConfig, but it requires getInitialProps which I do not use anymore. Any other ideas how to solve this?

**Comment by michrome** [comment, marked as answer]
When you promote an app on Heroku, you’re creating a [release](https://devcenter.heroku.com/articles/platform-api-reference#release) which is combination of the [slug](https://devcenter.heroku.com/articles/slug-compiler) from “app A” (e.g. dev) and the environment from “app B” (e.g. staging). Because you’re re-using the slug, you’re **skipping the build process by design**.

For anything that’s running in the Node.js environment (e.g. API routes or `getInitialProps`), you’ll get app B’s environment. But for anything you baked in during the build (e.g. `getStaticProps`), you’ll get app A’s environment.

So I think your choice here is:

* Move back to using `getInitialProps`. Despite recommendations to use `getStaticProps` where possible, `getInitialProps` is not a deprecated method and this is perfectly valid reason to use it.
* Use the [Heroku Platform API](https://devcenter.heroku.com/categories/platform-api) to create your own your own promotion workflow. You need a [build](https://devcenter.heroku.com/articles/platform-api-reference#build) in that process to create a new slug rather than reusing the same slug. You’re giving up having identical dev and staging slugs but, ultimately, I think that’s what you’re trying to achieve.
